modules/secure-preview/secure-preview.php

```diff
@@ -0,0 +1,150 @@
+			'module_url' => $this->module_url,
+			'slug'       => 'secure_preview',
+			'autoload'   => true,
+		] );
+	}
+
+	/**
+	 * Initialize secure preview module
+	 */
+	public function init() {
+		SecurePreviewEndpoint::init();
+
+		// Load block editor JS
+		add_action( 'enqueue_block_editor_assets', [ $this, 'load_block_editor_scripts' ], 9 /* Load before custom status module */ );
+
+		// Load block editor CSS
+		add_action( 'enqueue_block_editor_assets', [ $this, 'load_block_editor_styles' ] );
+
+		// Preview rendering
+		add_filter( 'query_vars', [ $this, 'add_preview_query_vars' ] );
+		add_action( 'posts_results', [ $this, 'allow_secure_preview_results' ], 10, 2 );
+	}
+
+	public function load_block_editor_scripts() {
+		$asset_file = include VIP_WORKFLOW_ROOT . '/dist/modules/secure-preview/secure-preview.asset.php';
+		wp_enqueue_script( 'vip-workflow-secure-preview-script', VIP_WORKFLOW_URL . 'dist/modules/secure-preview/secure-preview.js', $asset_file['dependencies'], $asset_file['version'], true );
+
+		$generate_preview_url = '';
+		$post_id              = get_the_ID();
+
+		if ( $post_id ) {
+			$generate_preview_url = SecurePreviewEndpoint::get_url( $post_id );
+		}
+
+		$custom_status_module = VIP_Workflow::instance()->custom_status;
+		$custom_status_slugs  = wp_list_pluck( $custom_status_module->get_custom_statuses(), 'slug' );
+		$custom_post_types    = $custom_status_module->get_post_types_for_module();
+
+		wp_localize_script( 'vip-workflow-secure-preview-script', 'VW_SECURE_PREVIEW', [
+			'custom_post_types'    => $custom_post_types,
+			'custom_status_slugs'  => $custom_status_slugs,
+			'expiration_options'   => $this->get_link_expiration_options(),
+			'text_preview_error'   => __( 'There was an error generating a preview link:', 'vip-workflow' ),
+			'url_generate_preview' => $generate_preview_url,
+		] );
+	}
+
+	public function load_block_editor_styles() {
+		$asset_file = include VIP_WORKFLOW_ROOT . '/dist/modules/secure-preview/secure-preview.asset.php';
+
+		wp_enqueue_style( 'vip-workflow-secure-preview-styles', VIP_WORKFLOW_URL . 'dist/modules/secure-preview/secure-preview.css', [ 'wp-components' ], $asset_file['version'] );
+	}
+
+	public function add_preview_query_vars( $query_vars ) {
+		$query_vars[] = 'vw-token';
+
+		return $query_vars;
+	}
+
+	public function allow_secure_preview_results( $posts, &$query ) {
+		$token = $query->query_vars['vw-token'] ?? false;
+
+		// If there's no token, go back to result processing quickly
+		if ( false === $token ) {
+			return $posts;
+		}
+
+		// Only allow secure preview on individual post queries
+		$is_preview = $query->is_preview() && 1 === count( $posts );
+
+		if ( ! $is_preview ) {
+			return $posts;
+		}
+
+		if ( Token::validate_token( $token, $posts[0]->ID ) ) {
+			// Temporarily set post_status to 'publish' to stop WP_Query->get_posts() from clearing out
+			// unpublished posts before render
+			$posts[0]->post_status = 'publish';
```

Comment:
Where does the correct status get restored? This feels risky, since other code (and other hooks) may run before it is restored and incorrectly assume the post is published.

Reply:
Would be curious about this as well.

Reply:
This is a good concern! I looked through the query code and found that after [the `posts_results` filter](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3401) is called and [the logic that allows previewing](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3452-L3484) is done executing, [`the_posts` filter is called](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3558). I added some code to revert the post status using `the_posts`:

https://github.com/Automattic/vip-workflow-plugin/blob/8cf9f782e01f6a174614e306b3a4a2a2994bbe02/modules/preview/preview.php#L101-L124

Most of the other logic in `WP_Query::get_posts()` between `post_results` and `the_posts` is skipped during a preview, so there's not much opportunity to get unexpected side-effects from other parts of WordPress or plugins.